src/diamonds/facets/SnapshotsLibrary.sol

```diff
@@ -102,60 +104,102 @@ library SnapshotsLibrary {
-                "snapshot heights too close in MadNet"
-            );
-
+            ethBlocksSinceLastSnapshot = block.number - previousSnapshot.ethHeight;
+            
+            if (cs.epoch > 1) {
+
+                require(
+                    !previousSnapshot.saved || ethBlocksSinceLastSnapshot >= ss.minEthSnapshotSize,
+                    "snapshot heights too close in Ethereum"
+                );
+
+                require(
+                    !previousSnapshot.saved || madHeight - previousSnapshot.madHeight >= ss.minMadSnapshotSize,
+                    "snapshot heights too close in MadNet"
+                );
+            }
         }
 
+        // Lock rewards
         ParticipantsLibrary.ParticipantsStorage storage ps = ParticipantsLibrary.participantsStorage();
-        StakingLibrary.StakingStorage storage stakingS = StakingLibrary.stakingStorage();
-
-        for (uint idx=0; idx<ps.validators.length; idx++) {
-            if (msg.sender==ps.validators[idx]) {
-                StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount + stakingS.rewardBonus, cs.epoch+2);
-            } else {
-                StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount, cs.epoch+2);
+        {
+            StakingLibrary.StakingStorage storage stakingS = StakingLibrary.stakingStorage();
+            for (uint idx=0; idx<ps.validators.length; idx++) {
+                if (msg.sender==ps.validators[idx]) {
+                    StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount + stakingS.rewardBonus, cs.epoch+2);
+                } else {
+                    StakingLibrary.lockRewardFor(ps.validators[idx], stakingS.rewardAmount, cs.epoch+2);
+                }
             }
         }
 
+        // Check if sender is the elected validator allowed to make the snapshot
+        require(
+            mayValidatorSnapshot(
+                int(ps.ethdkgDiamond.numberOfRegistrations()),
+                int(ps.ethdkgDiamond.validatorIndex(msg.sender)),
+                int(ethBlocksSinceLastSnapshot) - int(ss.snapshotDesperationDelay),
+                keccak256(_bclaims),
+                int(ss.snapshotDesperationFactor)
+            ),
+            "validator not among allowed"
+        );
+
+        // determine whether validators changed and thus ethDKG needs to be rerun
         bool reinitEthdkg;
         if (ss.validatorsChanged) {
             reinitEthdkg = true;
         }
         ss.validatorsChanged = false;
 
-        emit SnapshotTaken(chainId, cs.epoch, height, msg.sender, ss.validatorsChanged);
+        emit SnapshotTaken(chainId, cs.epoch, madHeight, msg.sender, ss.validatorsChanged);
 
         cs.epoch++;
 
         return reinitEthdkg;
     }
 
+    function mayValidatorSnapshot(int numValidators, int myIdx, int blocksSinceDesperation, bytes32 blsig, int desperationFactor) internal pure returns (bool) {        
```

Comment:
Do the input parameters need to be `int`? numValidators, myIdx, blocksSinceDesperation or desperationFactor can have negative values?